### PR TITLE
allow non hosts to change factions later.

### PR DESCRIPTION
**Pull request by gfgtdf**
previously players had to leave an rejoin the game to change their faction.
this was most annoying in coop games where you want might want to choose
your faction in consultation with the other players. Also you previously could
not even checkout the different available factions again after you joined
the game.

### CODE REVIEW

File: src/gui/dialogs/multiplayer/mp_join_game.cpp [hunk 1]
```diff
@@ -154,6 +154,7 @@ bool mp_join_game::fetch_game_config()
 
 	int side_num = 0, nb_sides = 0;
 	for(const config& side : get_scenario().child_range("side")) {
+		// TODO: it can happen that the scenario specfies that the controller of on side should also gain control over another wise.
```
Comment (stevecotton) [inline src/gui/dialogs/multiplayer/mp_join_game.cpp]: "wise" seems a typo

File: src/gui/dialogs/multiplayer/mp_join_game.cpp [hunk 2]
```diff
@@ -307,6 +308,58 @@ void mp_join_game::pre_show(window& window)
...
 
+void mp_join_game::show_flg_select(window& /*window*/, int side_num)
+{	
+	if(config& side_choice = get_scenario().child("side", side_num -1)) {
+		if(!side_choice["allow_changes"].to_bool(true)) {
+			return;
+		}
+		const config& era = level_.child("era");
+		if(!era) {
+			ERR_MP << "no era infomration\n";
```
Comment (soliton-) [inline src/gui/dialogs/multiplayer/mp_join_game.cpp]: information

**Comment by Vultraz** [comment]
I have some feedback, will get to this soon.

**Comment by gfgtdf** [comment]
ok fixed the types mentioned

> I have some feedback, will get to this soon.

if so please hurry a bit, is been already a week now.

**Comment by Vultraz** [comment]
Ok so looking at this again, turns out something I thought was done incorrectly is actually fine... I guess really the only feedback I'd have is to reduce the code duplication with `fetch_game_config`. You could definitely move the Faction Select invocation to its own function.

**Comment by gfgtdf** [comment]
ok now i have a new problem: the dialog might delete/recreate the change-faction buttons while the change-faction dialog is up which is bad since we are currently executing the onclick handler for exaxtly that button which might lead to crashes.

**Comment by Vultraz** [comment]
Yeah, I've come across that problem in a different context. Dunno how to solve it, really...

**Comment by gfgtdf** [comment]
i think i need to do something like in the click handler just fire an (sdl) event and do the real action then in the handler of that event. Not sure on the details yet.

**Comment by Vultraz** [comment]
TBH if this becomes too much trouble it might be worth just dropping it. I'm not sure we want to start getting into implementing callback handling exceptions in the main SDL event loop.

**Comment by gfgtdf** [comment]
> I'm not sure we want to start getting into implementing callback handling exceptions in the main SDL event loop.

hmm i don't think we need a special handling for exceptions here, it's still singetreaded and whatever is thrown will go to whatever calls pump() (in 1.14 not sure how it works on master) (just like it also happens for 'normal' exceptions i think)

**Comment by Vultraz** [comment]
I meant exceptions as in special cases not C++ exceptions.

**Comment by CelticMinstrel** [comment]
Why not just prevent the dialog from updating (and deleting the buttons) until the FLG dialog returns? This could be as simple as setting a flag.

**Comment by gfgtdf** [comment]
> Why not just prevent the dialog from updating (and deleting the buttons) until the FLG dialog returns? 

not proceccing input would be a rather easy option but i'd prefer if the flg dialog would automatically closes when the game starts. Proceccing nework data but not updating the dialog is somewhat nontrivial becasue i'd need to update the dialog then after that and 'after that' is still inside the click handler.

**Comment by stevecotton** [comment]
Maybe the client should send a message to the host that the user is changing their faction, so that the host doesn't start the game while the user is choosing?

However, with 1.14.0, 1.14.1 or 1.14.2 (without this PR being involved) I've already seen a similar crash if the host opens the faction dialogue for a unowned side, and then a new player joins and takes that side.

**Comment by gfgtdf** [comment]
I am not 100% sure what type of message you mean, but such a message cannot 'fix' such a crash because it also need to work when the hosts uses an older client that doesn't understand such messages. But I also don't like the idea of blocking the host from starting the game in general, a visual notification to the host would be nice but not a high priority for me currently.

> I've already seen a similar crash...

hmm ye will treat that aswell.

**Comment by stevecotton** [comment]
The problem that the Codacy Bot keeps complaining about has a one-word fix:
```
+++ b/src/utils/scope_exit.hpp
@@ -24,7 +24,7 @@ class scope_exit {
        std::function<void ()> f_;
 public:
        template<typename F>
-       scope_exit(F&& f) : f_(f) {}
+       explicit scope_exit(F&& f) : f_(f) {}
        ~scope_exit() { if(f_) { f_(); }}
 };
 } // namespace utils
```

About blocking the host from starting the game, that already happens if the player hasn't chosen a side yet, so it's not a new behaviour.

**Comment by gfgtdf** [comment]
ok it finally works. The gui2 timer class cannto handle mutipel timer executing at the same time that well so i cannot remove the timer from inside its own callback.

> The problem that the Codacy Bot keeps complaining about has a one-word fix:

ye making codacy happy was not really priority so i wanted to do that one last, changed it now .

> About blocking the host from starting the game, that already happens if the player hasn't chosen a side yet, so it's not a new behaviour.

Hmm well we could change it guess are but i probably still won't spend time on it for now.

**Comment by gfgtdf** [comment]
> However, with 1.14.0, 1.14.1 or 1.14.2 (without this PR being involved) I've already seen a similar crash if the host opens the faction dialogue for a unowned side, and then a new player joins and takes that side.

i tried usingg 1.14.3 official windows release and could not reproduce don't have older versions installed.

**Pull request merged**